log4j-api/src/main/java/org/apache/logging/log4j/status/StatusLogger.java

```diff
@@ -16,311 +16,567 @@
+    /**
+     *
+     * Sets the level of the default listener.
+     *
+     * @param level a level
+     * @deprecated Instead either use {@link #setDefaultListenerLevel(Level)}, or {@link #getListeners() get the specific listener} and change its {@link StatusListener#getStatusLevel() level}.
+     */
+    @Deprecated
+    public void updateListenerLevel(final Level level) {
+        requireNonNull(level, "level");
+        setDefaultListenerLevel(level);
     }
 
     /**
-     * Returns a thread safe Iterable for the StatusListener.
+     * Returns the listener collection.
      *
-     * @return An Iterable for the list of StatusListeners.
+     * @return a thread-safe read-only collection of listeners
      */
     public Iterable<StatusListener> getListeners() {
-        return listeners;
+        listenerReadLock.lock();
+        try {
+            return Collections.unmodifiableCollection(listeners);
+        } finally {
+            listenerReadLock.unlock();
+        }
     }
 
     /**
-     * Clears the list of status events and listeners.
+     * Clears the event buffer and listeners.
      */
     public void reset() {
-        listenersLock.writeLock().lock();
+        listenerWriteLock.lock();
         try {
-            for (final StatusListener listener : listeners) {
-                closeSilently(listener);
+            Iterator<StatusListener> listenerIterator = listeners.iterator();
+            while (listenerIterator.hasNext()) {
+                StatusListener listener = listenerIterator.next();
+                closeListenerSafely(listener);
+                listenerIterator.remove();
             }
         } finally {
-            listeners.clear();
-            listenersLock.writeLock().unlock();
-            // note this should certainly come after the unlock to avoid unnecessary nested locking
-            clear();
+            listenerWriteLock.unlock();
         }
+        buffer.clear();
     }
 
-    private static void closeSilently(final Closeable resource) {
+    private static void closeListenerSafely(final StatusListener listener) {
         try {
-            resource.close();
-        } catch (final IOException ignored) {
-            // ignored
+            listener.close();
+        } catch (final IOException error) {
+            final String message = String.format("failed closing listener: %s", listener);
+            new RuntimeException(message, error).printStackTrace(System.err);
         }
     }
 
     /**
-     * Returns a List of all events as StatusData objects.
+     * Returns buffered events.
      *
-     * @return The list of StatusData objects.
+     * @return a thread-safe read-only collection of buffered events
      */
     public List<StatusData> getStatusData() {
-        msgLock.lock();
-        try {
-            return new ArrayList<>(messages);
-        } finally {
-            msgLock.unlock();
-        }
+        return Collections.unmodifiableList(new ArrayList<>(buffer));
```

Comment:
Do we need to prevent modification of the returned data, since it is a copy?

A typical application for us would be in tests where we can:

- get a snapshot of the data,
- assert the presence of some `WARN` messages,
- assert that there are no other `WARN` messages.

Reply:
I like the idea of making the returned collection read-only. This makes it clear that mutations on the returned buffer will not be propagated.

Reply:
Usually, an unmodifiable list view is made on the modifiable list. If you're returning an immutable copy, at least in newer Java, you'd just use `List::copyOf`. If that's the intention here, then go for it; otherwise, returning an `ArrayList` copy is sufficient since the caller can't change the `StatusLogger` collection.

Reply:
Documented the rationale in c4c941e9125cf259ad36e24facd91c046bf0ad67.